# Lab 4: RSA

Write <span style="color:red">HACKOOLIQUES HERE</span>

## A) Using RSA

Check that you understand the arithmetic involved in the operation of the RSA cipher by encrypting a message of your choice with the following 3072-bit modulus and public encryption key:

In [1]:
n = 0x1bc157802f2fe3faa1df716cbba591dd142070f07f3eccaea8c53dfc954df81795c6f19e530bd8bfc846548a699ff5d98e305bf4b9d30d188d51b49c7c9b759ecc11b4ed27856afab38e9777a0c267d4946694afd6a49b3a396b9508c2e7aa088813124e23d3308de22885ebe024837483e82c5ff913a94b72d48541d68e5dbd2654ae1abdb02f82e260707bc63240bfe1cc80c6a5790dc70df422fc5298786a9f4e59446c3b0221ae967a70536c0038d9849561bd288e3fd5a37a9d834a2f42a7b2c0e547359b2c4616f8e758f5fc6d458cd5d18760043f49b3ef625ec97d79595b8d1bc036cb734879b621f4ddfb33d82c42a3b4c4f0d8648d9006a53f9a330ac4d49bc12fd540aaa879de36fc780b38a916fc5ff0e9302481733b2280b199819cb61087a8c5c55ec09f3041621f01c0cf08180b87c1828f7f8c9b84447aa691ea968eb871727d1df94e325d240150c14c205bf7290d1aaf4299d2cc6dbcab0da759391c2bc60eb21e04996336af0895e5a7d68e2ed12cebe7d031267744097

In [2]:
e = 0x4e0c390c474749a26cef641ffeb05b8fff94e84e45638dc18daac67aab3f043ff8bec60659aa5cebc1ecbe43970112c98a76ec3ee4ee3dbc702592eb61c0dcf6c38af806f7b10d982e84391d09991e760c6621e06afe2e1e257aec802630b48ef8b24e4a6495d9bacef8d21ad251440ceefbdd1184be2e115c397a84f7527ce3b9b789055acf22157b947b6872c8ae7eb497dfabf81edfd69cb0d3a082721b22fea5125f206aa1710b850809891ef94818e02078d9f6e6776d3fc976d366a15252b109315443bd9a202f4e14092079038250af9627fa066a11f85c5e4f70891e79a8bdaab043c8a17ca4e6b8e80ca5e0cff02e5b20f4b342f6ea90b20255c7dc666071e87363de6e862cd83a27f699e58f32347d266b51e213688f7e96dc6305bd53001d557c0965f71cbe50cb94b216a5b45e815eb2c81f6167ed28785fb53792581760a60070e6a80bb733d172a145b899dc25cc754eff12b143007642063e0df97b03005c46155d873de324724c37a07670b1b3cda1c3eec6fbec82f0261b

I would like you to write your own modular exponentiation function to really appreciate how fast this operation is once it is performed properly (_i.e._, implement your own version of Python's 3-argument `pow` for nonnegative exponents).

To check your implementation, here is the encryption of the byte array ``b"Test vector"``, viewed as an integer written in base 256 (most significant byte first):

In [10]:
expected = 0x435b19ade1a782a84bd0d17cd317252b928ee93c21e06bbd4e4ba93c7e2ed1a987881f82802d4db16ffd480696971a10abad5dfc027f56f7fa06a11a25b4fab69b28c3db9fd05cf342710a440244fba44925dd9e3006d7533aaca9cffd09e0c43ee67e80aca077073092503a17cbe7631aca4a4c65e4d5e8ba2933e34ecec89aadb2ede331fd57d227e75156c5760e8cb207774870ad51b9fd6895ee8565ebbafe20b5e3914aa8c8daaee2fefc715cc9f3d7c4c5cc8cb1e0d9074afc6c08fd3bd19a279573cb90acd7891630da1a29ff1a6345b3b07b5f5a16bc0edf6275f8dd9e4e7964011732e1b5fef670c639f84d7594619277aafb1f7238be8d11185076b35b5d8667b7bae7937851c1d01198009d6e94c2c25831f15368e7afb2f8e8da6df781d68e9821a9a8c95d15e106a77cb090a59bc0c261c4e56af031a606a02ce450da73e8671d44049d87b3b4e54f3c610fa01d4317e6c9e51f04b141e774bee5ede55261d8c65947e2d4ec0055549aff59e5eef3d3ad200478501f514b2443

In [ ]:
def exp(n,e,mod=10):
    res=1
    n %= mod
    while e>0:
        e,r = divmod(e,2)
        if r == 1 : res = (res * n) % mod
        n = (n * n) % mod
    return res

In [18]:
message = b"Test vector"
m = int.from_bytes(message, byteorder="big")
result = exp(m, e, n)
print( hex(result) == hex(expected) )

True


Utiliser l'exponentiation rapide permet de calculer en O(log n) plutôt qu'en O(n) avec le calcul naïf

## B) Exploiting RSA malleability

One of the weaknesses of unpadded RSA is the fact that an attacker can easily manipulate a ciphertext to produce a predictable effect on the plaintext. Demonstrate this by performing the following: the number $c$ is the encrypted version of some secret integer $m$; produce the ciphertext $c'$ that corresponds to the integer $256 \cdot m$ so that Bob, when decrypting $c'$, will get an extra `00` byte added at the end of $m$.

Explain (in a text cell) why your attack works! (NB: Since you have the encryption key, you can easily produce the ciphertexts corresponding to any plaintext of your liking to test your attack on messages of your own.)

In [40]:
c = 0x12fb9fd9cb1fd608d19ec3fb5dda78cfc57811c7e6ba6f9dd93482957824caa94a291299aaff3c0ac46443f8d5d27b4e1958f288d861a046c4871bd2984ac6b5dc17fed10c5e810d683403bf3556d5cf8d5af47e05613d0b1a02956147f2d694086df7504164904ea6825de828c55af1c6b06456bca7bc99617b3d28852298e7654e4d04ba06c95cdb6669a66aba4ced243d9a35e778168e972ba62546f03487a0c7125a86ed5968f715527b411bf6249cfff97610f4f761e8b476bdea2d21b4db21a761dc9d28b0d03e6ac3c099af3c1e2e0e52ae8593e3dc6d1adae2810d06e7800c35aa5ac8350acc407525bef08f10622bada60fcd24803fcdd41f6e1a439384d4a5b9e067b48e09e864cd6b0347f3190475764f423f71e757fa41d26c739fc80ab64168865fdcd1ca6663008aaa53b370f6688b1828b0eb33e6c974be35b6d01adaff50810668d77bbb4f3abf06e3b624ac7e6f74176bed39d702a8ee0334f598c338e1d412afbb012a920eba0b69abcc222beee35b23d2b425dc22d489c

En binaire, multiplier un nombre par 256 correspond à un décalage d'1 octet vers la gauche, ce qui ajoute Ox00 comme octet de poids faible.
Ajoute (256 ^ e) % n au chiffré revient à faire ( (m * 256) ^ e ) % n.
En déchiffrant le chiffré altéré, le message original sera donc décalé d'1 octet vers la gauche.

In [48]:
message = b"Je suis un hackeur hors norme"
m = int.from_bytes(message, byteorder="big")
m_prime = m * 256   # ajoute un octet 0x00 à la fin

c_ = exp(m, e, n)
c_mprime = exp(m_prime, e, n)

k = exp(256, e, n)              # 256^e mod n
c_from_c = (k * c_) % n
c_256 = (c_ * 256) % n

print( c_from_c == c_mprime)

True


## C) Breaking RSA

Note that the above modulus $n$ would be state-of-the-art hard to factor without further information. Suppose, however that I was careless enough to reuse for $n$ a modulus for which you know a pair of public and private keys $(e',d')$:

In [49]:
ee = 0x197cbcd9c3b22d4db9ef8f060b724cf3c2d3e85faa76ef87f0f18b3c09f577ffcbad82791d6656f297c3d4b5b9b6f4ad101d53713a530063e7f2903bd358d13481074e6c7e1eb3a7562ec500222d96abacfda8f88989d8f4158933ca08f2efb310d744b9c675b8705031c9f398a4bed41cf125bf3a985f3d4c9711ecc776f25a7f973b36c74802ed561dab96e1d5c320e4cc7916b5f26b85886df3f013fd4f744faee5b68d82fe5628f9dbd6bb2582e22659c804678101e1a080a6bc8d019213555e1e74e533c08c7b804b0efc12c0fb7688c292a781c1cb85bb35f13469726a9dba06b7455ad97883ebfa4358e2ccd836cc3730b4b64e14173598daeb4eb047f1e37f65231bf11833c39429b10b16f342c0b17f9c3d0c627df9c6f800bc0ffd87e53caeb5b6e7fb96b3b3799edbcdc5ca2b41ba2a549ecea59b9033d06818c50b762b3971f233b3298f4cc32d10518f2084a903c9ea3e76ab1dcbf7eadf1e823a1ee9738df01392166dc6854c2a8042cd177a525dd3d63d77db83745d9d68e93

dd = 0x120634530f7b4b60f4bb6e9ef1ac275879e84d6f0aa72958c7a75c6407e79b374d9a48a4b29127ce3aabd5043269f4a3b3a29a997ab8aa66320a08074c2dda0b3458a6394717f035f4dd17ce057f2149c891a1c5afca948b7030e03a81e1dd79ccf6ebd7be882b3e10fa6e19de6b0de10f6dc7ffe30c504b5bb46a5fad6b706d57e0be20eda1e6d6de2cd724dfb8a3a72d59ed1285f744c15972efd5ae671a4324fc3838a558997e2fb75ff658300767a0a177ff968fed74d34c1c4cb40b520e91b718baa1b74c007ce86c0727b5239dec41e4ea32ffeaf8e51124c9845df91d7746cfa186bf06cfab25ea439b2f212b7d8335ba20a1310715f29bc1f67301ad852b33fd72cf7e544d89e8841b12d4b821fe32d2e6fcb4de6412fb370e59c46088fc5da7794e0cbab6ed337a4680dbd8d81412c79f1744059de7681de0128d936fd9a6bb6482582cf791923b5340974f6ce59e112fba6ca9fa60488fb1a5bd7b3dc8d458cb21f2665bb1c23aa6a376c2fd9e97cae388f35690ea58e0cbe7557b3

Using the knowledge of $e'$ and $d'$, launch a Fermat attack against $n$ in order to recover its prime factors $p$ and $q$.

In [ ]:
import math, random

def factor_from_ed(e, d, n, tries=50):

    k = (e * d) - 1
    if k % 2 == 1:
        raise ValueError("k doit être pair")
    
    s = 0
    r = k
    while r % 2 == 0:
        r //= 2
        s += 1

    for attempt in range(tries):
        g = random.randrange(2, n-1)
        y = pow(g, r, n)
        if y == 1 or y == n-1:
            continue
        for j in range(s):
            x = pow(y, 2**j, n)
            if x == 1:
                p = math.gcd(y-1, n)
                if 1 < p < n:
                    q = n // p
                    return p, q
                break
            if x == n-1:
                break
    raise RuntimeError("Échec : augmenter 'tries' ou vérifier les entrées")

In [64]:
p,q=factor_from_ed(ee, dd, n)
print("p", hex(p))
print("q", hex(q))
print("n", p*q == n )

p 0x111628ffc511fd8733b495b75f0cba471b32fd9d2127a0e531c022375bee8c911370c368f41d96673e996930f1304feb599a4234773e5f22c37c8a15c41a8b574acf307f9126b69e2a9aae09c6eff12ab4a896ff60853e7206134adc85c97e2009fda2005cb3f1bcd674576844e45b954fd51783c0dbce1c2041f3ef2e58c751bc8db9e78e6e70289a0720343af72622232fb5a7b202910f94f01ffd58b5f1c9c47b5a560205c4197e7cd5cd271b5022bbf6b3d11251039a903f68ec8d3cfb7bb
q 0x19fd80516f2d179caac721ddd4b7387085026971d2c8af69b3f9e8a6fe5e7a55e1febd90fdbe1b789d7da334a26dfbf55b0e3af241aa00aa934cecce3e6a8817d8d82ca9c626e2b086cca61a2bf329f8f09fb382286339dd67e45bf7bc7839aa0e1c4ab6a0ccf51c6e3b1641d6190ef48b301398a3bf0f32a5ce3174208a490f381864dfac26025792306a0828b983206ce9290e79043ff383b4a6ace9064fc330612d6df37f4271c137d0aa73187abdc8c35e9f85225aa10cf85c4a235d606d5
n True


Now knowing $p$ and $q$, you should be able to easily recover the private key $d$ corresponding to the $e$ of part A) and use it to find the decrypted messages $m$ and $m'$ corresponding to $c$ and $c'$ of part B).

In [79]:
def decypher(n,e,p,q,c):
    return pow(c, calc_d(e,p,q) ,n)

def calc_phi(p,q):
    return (p-1)*(q-1)

def calc_d(e,p,q):
    return pow(e,-1, calc_phi(p,q))


In [80]:
m_int = decypher(n,e,p,q,c)
m_original = m_int.to_bytes((m_int.bit_length() + 7) // 8, "big")
print(m_original)

b'It works!'


## D) Generating keys

Generating fresh parameters $(n,e,d)$ to use in RSA is, however, quite simple (_i.e._, reasonably fast). Help Alice and Bob by generating

- two 1536-bit prime factors $p$ and $q$;
- the corresponding RSA modulus $n$;
- a public exponent $e$ of your liking;
- the corresponding private exponent $d$.

For the first step, you can use the provided `is_prime()` function to quickly test if a given integer is (with large probability) prime. How long does generating RSA keys take on your machine? You can compare the performance of your key generation algorithm with that of a standard library, _e.g._ [OpenSSL](https://docs.openssl.org/3.1/man1/openssl-genrsa/).

In [82]:
import random
from is_prime import is_prime

def generate_prime(bits):
    while True:
        p = random.getrandbits(bits) | 1  # s'assurer que p est impair
        if is_prime(p):
            return p
        
def generate_rsa(e, bits):
    p, q = generate_prime(bits), generate_prime(bits)
    n = p * q
    d = calc_d(e,p,q)
    return n, d

In [83]:
bits = 1536
e = 65537
n, d = generate_rsa(e, bits)

print("n =", n)
print("e =", e)
print("d =", d)


n = 4093108676848011069086932372097502156397296906565023665178942000588233402995319067787770385552179380684204624425380048829675566521287499667452853389119633418967329243135454449911684171104920622816732737378399593057913325550253011864391146447256581778169114035971529497359836339474943081308174186083332154051121227330969039693800477470474841806838493750130118157660620056172765317788424783849645776840867461690542914242639025564437453560392385229518448236377406361454389288909078226517619180822260310505427134554963948210407734733008058082724832556952250944773155491054588942336724835119679499939272399869703805666988266905213218710867275997685571282430659671748440273987170547144333573129993902323369017470051782026324254015621367913651074054591503793951390422316862936891034605292151320265237505398133925250658932517993028567852726552755592770749262061392476972226559767517309134687837377303686661611690705292670453481452341
e = 65537
d = 12180584787916254921708720578149378909046261130161230532

In [84]:
n = 4093108676848011069086932372097502156397296906565023665178942000588233402995319067787770385552179380684204624425380048829675566521287499667452853389119633418967329243135454449911684171104920622816732737378399593057913325550253011864391146447256581778169114035971529497359836339474943081308174186083332154051121227330969039693800477470474841806838493750130118157660620056172765317788424783849645776840867461690542914242639025564437453560392385229518448236377406361454389288909078226517619180822260310505427134554963948210407734733008058082724832556952250944773155491054588942336724835119679499939272399869703805666988266905213218710867275997685571282430659671748440273987170547144333573129993902323369017470051782026324254015621367913651074054591503793951390422316862936891034605292151320265237505398133925250658932517993028567852726552755592770749262061392476972226559767517309134687837377303686661611690705292670453481452341
e = 65537
d = 1218058478791625492170872057814937890904626113016123053267389502685083480455585513207270485823643963890383184921009309128052284570008851580242199057753638564019100999875959658462053136229294397161828258496588602826014062105476059178650541360770940299672432840129281776508062439977109341513241697837606649685811332478384564156860868091408987896284101249809843209619834184590955368613571700862408129540983537625320940178436439195923274742333837208466946853747784552049892781546790850968645643067536414034001082943647487452313353105216700517612936346766293829615551658264622001178019312688599164175694373229259148927234566740308350952870418099993596356671776390837337400267953171564651319859710230855379570824909291637279948488003982496349715508637556810639518139206844316119344839000147128767785725861476538522115903140236957430039981678858571214580469049752298529387941856564776864680214203135383781930139598711501998814077173

In [91]:
def rsa_encrypt(message: bytes, n, e):
    m = int.from_bytes(message, "big")
    c = pow(m, e, n)
    return c

def rsa_decrypt(ciphertext: int, n, d):
    m = pow(ciphertext, d, n)
    return m.to_bytes((m.bit_length() + 7) // 8, "big")

In [93]:
message = b"Hello Bob, ici Alice !"  
print("Alice veut envoyer à Bob :", message)

ciphertext = rsa_encrypt(message, n, e)
print("Alice chiffre le message avec la clef publique de Bob :", hex(ciphertext))

decrypted = rsa_decrypt(ciphertext, n, d)
print("Bob déchiffre le message reçu avec sa clef privée :", decrypted)

Alice veut envoyer à Bob : b'Hello Bob, ici Alice !'
Alice chiffre le message avec la clef publique de Bob : 0xae20c246a986e662df922016693335556591cadc207a95936ae8ee35b70810b69ffe50a28926b54b01849fc08dfec0441e2c195213b6752fd7959cc1c9fb99841d08bd0da59186ffb867f469d004008acda9f745ebd4ef1adc06309213218ab96c74db7d7d0927fc53a369cdc2ff735f22695cf1a71ad077c73dd5113ef7d354e3eb6462117dc518a0981a62864188eb2fb57f11799235756b52f75940743a6fd35f88b56c5270b9b885ba29f8eec552fa4da1fa71da7edeccbbe308d44d1bc748976d3f2f88e1efafddef368bffe8cad4a0bdd406e8c64e436b8ab4b2e2d1f45f726e903e25c237b80119a1740c322c2f4d23cd1f2e333de5c9c118d028a117d24a79d13637b7768910f8c2e313563296f899fd0d44f224cb4eb910abcf56cc17ed7c12bb22470370e8cbd251e40b1b5cdb19f16c5c402d2e4e595b53828b5099754301606502338dab8a332fe6f6f391a32c264bb8b1bd428f1b587f387adc53fe30eafd8f5b8a9b04b10a4089f01ef067f2b0966eca6b2f5d02af5ad78f4e
Bob déchiffre le message reçu avec sa clef privée : b'Hello Bob, ici Alice !'


In [85]:
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.backends import default_backend

def generate_rsa_hazmat(bits=1536, e=65537):
    # Génère la clé privée RSA
    private_key = rsa.generate_private_key(
        public_exponent=e,
        key_size=bits * 2,        #  modulus ~2*1536 bits
        backend=default_backend()
    )
    
    numbers = private_key.private_numbers()
    p = numbers.p
    q = numbers.q
    d = numbers.d
    n = numbers.public_numbers.n
    e = numbers.public_numbers.e

    return n, e, d


In [90]:
from statistics import mean
from time import time

def benchmark(func, label, loops=10):
    times = []
    for _ in range(loops):
        t0 = time()
        func()
        times.append(time() - t0)
    print(f"\n{label}")
    print(f"  → min :  {min(times):.4f}s")
    print(f"  → max :  {max(times):.4f}s")
    print(f"  → mean : {mean(times):.4f}s")

bits = 1536
e = 65537

benchmark(lambda: generate_rsa_hazmat(bits, e), "Hazmat")
benchmark(lambda: generate_rsa(e, bits), "Maison")


Hazmat
  → min :  0.0323s
  → max :  0.2845s
  → mean : 0.1593s

Maison
  → min :  7.4864s
  → max :  41.6320s
  → mean : 20.3003s


Résultats du benchmarks sur 10 générations de paires de clef RSA:
| Méthode | Temps min | Temps max | Temps moyen |
|----------|-----------|-----------|-------------|
| **Hazmat** | 0.0323 s | 0.2845 s | 0.1593 s |
| **Maison** | 7.4864 s | 41.6320 s | 20.3003 s |
